qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -382,7 +382,7 @@ def _run(
 
             # If the synthesis changed the operation (i.e. it is not None), store the result.
             if synthesized is not None:
-                synthesized_nodes[node] = (synthesized, synthesized_context)
+                synthesized_nodes[node._node_id] = (synthesized, synthesized_context)
```

Comment:
This looks fine in this case since `synthesized_nodes` is only used with one DAG (and not e.g. shared by the recursive control flow block handling).

But this optimization is something we should be careful when applying in other places, since the semantics of uniqueness in keys shifts from being global to local within a specific DAG (which means that keys will clobber each other if node indices from more than a single DAG are used in the same map). This was responsible for a bug we had in the visualization code after the initial DAG port to Rust.